net/smartdns/Makefile

```diff
@@ -0,0 +1,59 @@
+#
+# Copyright (c) 2018-2019 Nick Peng (pymumu@gmail.com)
+# This is free software, licensed under the GNU General Public License v3.
+#
+
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=smartdns
+PKG_VERSION:=1.2019.28
+PKG_RELEASE:=1
+
+PKG_SOURCE_PROTO:=git
+PKG_SOURCE_URL:=https://www.github.com/pymumu/smartdns.git
+PKG_SOURCE_VERSION:=982002e836e486fb4e360bc10e84e7e7197caf46
+PKG_MIRROR_HASH:=ae889b521ecf114660ce5597af5c361f5970a7dcb75dadf61b938ba3d2baba54
+
+PKG_MAINTAINER:=Nick Peng <pymumu@gmail.com>
+PKG_LICENSE:=GPL-3.0-or-later
+PKG_LICENSE_FILES:=LICENSE
+
+PKG_BUILD_PARALLEL:=1
+
+include $(INCLUDE_DIR)/package.mk
+
+MAKE_VARS += VER=$(PKG_VERSION) 
+MAKE_PATH:=src
+
+define Package/smartdns
+  SECTION:=net
+  CATEGORY:=Network
+  TITLE:=smartdns server
+  DEPENDS:=+libpthread +libopenssl
```

Comment:
Does this compile without deprecated APIs enabled?

Reply:
mips-openwrt-linux-musl-gcc -Os -pipe -mno-branch-likely -mips32r2 -mtune=24kc -fno-caller-saves -fno-plt -fhonour-copts -Wno-error=unused-but-set-variable -Wno-error=unused-result -msoft-float -mips16 -minterlink-mips16 -iremap/media/usb0/develop/openwrt/openwrt/build_dir/target-mips_24kc_musl/smartdns-1.2019.28:smartdns-1.2019.28 -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro  -I/media/usb0/develop/openwrt/openwrt/staging_dir/target-mips_24kc_musl/usr/include -I/media/usb0/develop/openwrt/openwrt/staging_dir/target-mips_24kc_musl/include -I/media/usb0/develop/openwrt/openwrt/staging_dir/toolchain-mips_24kc_gcc-7.5.0_musl/usr/include -I/media/usb0/develop/openwrt/openwrt/staging_dir/toolchain-mips_24kc_gcc-7.5.0_musl/include/fortify -I/media/usb0/develop/openwrt/openwrt/staging_dir/toolchain-mips_24kc_gcc-7.5.0_musl/include

compile options

the source is compiled with default openwrt gcc options.

Reply:
Sorry. I meant deprecated OpenSSL APIs.

Reply:
Not using any deprecated OpenSSL APIs, No compile warnings.

Reply:
I just tested this. It's full of deprecated API usage:

```
smartdns.c: In function '_smartdns_init_ssl':
smartdns.c:225:2: warning: implicit declaration of function 'SSL_load_error_strings'; did you mean 'ERR_lib_error_string'? [-Wimplicit-function-declaration]
  225 |  SSL_load_error_strings();
      |  ^~~~~~~~~~~~~~~~~~~~~~
      |  ERR_lib_error_string
smartdns.c:226:2: warning: implicit declaration of function 'SSL_library_init'; did you mean 'SSL_in_init'? [-Wimplicit-function-declaration]
  226 |  SSL_library_init();
      |  ^~~~~~~~~~~~~~~~
      |  SSL_in_init
smartdns.c:227:2: warning: implicit declaration of function 'OpenSSL_add_all_algorithms' [-Wimplicit-function-declaration]
  227 |  OpenSSL_add_all_algorithms();
      |  ^~~~~~~~~~~~~~~~~~~~~~~~~~
smartdns.c: In function '_smartdns_destroy_ssl':
smartdns.c:236:2: warning: implicit declaration of function 'ERR_free_strings'; did you mean 'ERR_load_strings'? [-Wimplicit-function-declaration]
  236 |  ERR_free_strings();
      |  ^~~~~~~~~~~~~~~~
      |  ERR_load_strings
smartdns.c:237:2: warning: implicit declaration of function 'EVP_cleanup'; did you mean 'EVP_PBE_cleanup'? [-Wimplicit-function-declaration]
  237 |  EVP_cleanup();
      |  ^~~~~~~~~~~
      |  EVP_PBE_cleanup

```
util.c:696:13: error: 'CRYPTO_LOCK' undeclared (first use in this function); did you mean 'CRYPTO_RWLOCK'?
  696 |  if (mode & CRYPTO_LOCK) {
      |             ^~~~~~~~~~~
      |             CRYPTO_RWLOCK
util.c:696:13: note: each undeclared identifier is reported only once for each function it appears in
In file included from util.c:29:
util.c: In function 'SSL_CRYPTO_thread_setup':
util.c:716:27: warning: implicit declaration of function 'CRYPTO_num_locks'; did you mean 'CRYPTO_realloc'? [-Wimplicit-function-declaration]
  716 |  lock_cs = OPENSSL_malloc(CRYPTO_num_locks() * sizeof(pthread_mutex_t));
      |                           ^~~~~~~~~~~~~~~~
util.c:731:2: warning: implicit declaration of function 'CRYPTO_set_id_callback'; did you mean 'BIO_set_callback'? [-Wimplicit-function-declaration]
  731 |  CRYPTO_set_id_callback(_pthreads_thread_id);
      |  ^~~~~~~~~~~~~~~~~~~~~~
      |  BIO_set_callback
util.c:732:2: warning: implicit declaration of function 'CRYPTO_set_locking_callback' [-Wimplicit-function-declaration]
  732 |  CRYPTO_set_locking_callback(_pthreads_locking_callback);
      |  ^~~~~~~~~~~~~~~~~~~~~~~~~~~
```

```

Reply:
```
diff --git a/src/smartdns.c b/src/smartdns.c
index a763840..c36a175 100644
--- a/src/smartdns.c
+++ b/src/smartdns.c
@@ -222,20 +222,22 @@ static int _smartdns_set_ecs_ip(void)
 
 static int _smartdns_init_ssl(void)
 {
+#if OPENSSL_API_COMPAT < 0x10100000L
 	SSL_load_error_strings();
 	SSL_library_init();
 	OpenSSL_add_all_algorithms();
 	SSL_CRYPTO_thread_setup();
-
+#endif
 	return 0;
 }
 
 static int _smartdns_destroy_ssl(void)
 {
+#if OPENSSL_API_COMPAT < 0x10100000L
 	SSL_CRYPTO_thread_cleanup();
 	ERR_free_strings();
 	EVP_cleanup();
-
+#endif
 	return 0;
 }
 
diff --git a/src/util.c b/src/util.c
index f26ec77..25ccba2 100644
--- a/src/util.c
+++ b/src/util.c
@@ -687,6 +687,7 @@ errout:
 	return -1;
 }
 
+#if OPENSSL_API_COMPAT < 0x10100000
 #define THREAD_STACK_SIZE (16 * 1024)
 static pthread_mutex_t *lock_cs;
 static long *lock_count;
@@ -743,6 +744,7 @@ void SSL_CRYPTO_thread_cleanup(void)
 	OPENSSL_free(lock_cs);
 	OPENSSL_free(lock_count);
 }
+#endif
 
 #define SERVER_NAME_LEN 256
 #define TLS_HEADER_LEN 5
@@ -965,4 +967,4 @@ int set_sock_keepalive(int fd, int keepidle, int keepinterval, int keepcnt)
 	setsockopt(fd, IPPROTO_TCP, TCP_KEEPIDLE, &keepcnt, sizeof(keepcnt));
 
 	return 0;
-}
\ No newline at end of file
**+}**
```

This codebase is nasty. I recommend running it through dos2unix to avoid CRLF problems.

Reply:
Thank you for reviewing the code, and the patch. 
All the code issues have been fixed, please review.